sandbox/cgroup/scanning.go

```diff
@@ -167,6 +159,45 @@ func PidsOfSnap(snapInstanceName string) (map[string][]int, error) {
 		}
 		return nil, err
 	}
+	return pathList, nil
+}
+
+// PidsOfSnap returns the association of security tags to PIDs.
+//
+// NOTE: This function returns a reliable result only if the refresh-app-awareness
+// feature was enabled before all processes related to the given snap were started.
+// If the feature wasn't always enabled then only service process are correctly
+// accounted for.
+//
+// The return value is a snapshot of the pids for a given snap, grouped by
+// security tag. The result may be immediately stale as processes fork and
+// exit.
+//
+// Importantly, if the per-snap lock is held while computing the set, then the
+// following guarantee is true: if a security tag is not among the results then
+// no such tag can come into existence while the lock is held.
+//
+// This can be used to classify the activity of a given snap into activity
+// classes, based on the nature of the security tags encountered.
+func PidsOfSnap(snapInstanceName string) (map[string][]int, error) {
+	// pidsByTag maps security tag to a list of pids.
+	pidsByTag := make(map[string][]int)
+
+	paths, err := InstancePathsOfSnap(snapInstanceName, 0)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, path := range paths {
+		pids, err := pidsInFile(path)
+		if err != nil {
+			continue
+		}
```

Comment:
I wonder if we should remove these lines and just add an empty pid list to the `pidsByTag` map; after all, a tag was found, so I think this can only happen if a pid has just exited.

My worry is that if right after InstancePathsOfSnap the only pid present in the only cgroup of the snap exited, and at the same time another pid was started in another cgroup of the same snap, we would return an empty `pidsByTag` map here. Though, maybe, this is not a problem -- it's still not clear to me how the result from this function is used.

Reply:
The problem is... what happens if we find a tag, but between `InstancePathsOfSnap()` and the loop, that snap exits... In that case, when it arrives to the loop, that file doesn't exist.

Anyway, this `PidsOfSnap()` function did exist before; what I did is to extract part of it (to the `InstancePathsOfSnap()` function) to reuse it and don't have the same code duplicated in two places.